### ISSUE TITLE
Recurrent x.neuron -> x connection gives IndexError

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo-loihi#152]
(On the master branch). Discovered while investigating #150 and trying to solve https://forum.nengo.ai/t/how-many-neurons-can-be-fully-connected/706.

```python
import numpy as np

import nengo
import nengo_loihi

n_neurons = 50

with nengo.Network() as model:
    x = nengo.Ensemble(n_neurons, 1)
    nengo.Connection(x.neurons, x, transform=np.ones((x.dimensions, x.n_neurons)))
    
with nengo_loihi.Simulator(model) as sim:
    sim.run(1.0)
```

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-35-7b9dd5e25fc9> in <module>
     15 
     16 with nengo_loihi.Simulator(model, precompute=True) as sim:
---> 17     sim.run(1.0)

~/CTN/nengo-loihi/nengo_loihi/simulator.py in run(self, time_in_seconds)
    558             logger.info("Running %s for %f seconds, or %d steps",
    559                         self.model.label, time_in_seconds, steps)
--> 560             self.run_steps(steps)
    561 
    562     def step(self):

~/CTN/nengo-loihi/nengo_loihi/simulator.py in run_steps(self, steps)
    720             self._make_run_steps()
    721         try:
--> 722             self._run_steps(steps)
    723         except Exception:
    724             if "loihi" in self.sims and self.sims["loihi"].use_snips:

~/CTN/nengo-loihi/nengo_loihi/loihi_cx.py in run_steps(self, steps)
   1077         """
   1078         for _ in range(steps):
-> 1079             self.step()
   1080 
   1081     def _filter_probe(self, cx_probe, data):

~/CTN/nengo-loihi/nengo_loihi/loihi_cx.py in step(self)
   1004 
   1005                     weights, indices = synapses.axon_weights_indices(
-> 1006                         spike.axon_id, atom=spike.atom)
   1007                     qb[0, cx_base + indices] += weights
   1008 

~/CTN/nengo-loihi/nengo_loihi/loihi_cx.py in axon_weights_indices(self, axon_idx, atom)
    451     def axon_weights_indices(self, axon_idx, atom=0):
    452         weight_idx = self.axon_weight_idx(axon_idx)
--> 453         w = self.weights[weight_idx]
    454         i = self.indices[weight_idx]
    455         return w[atom, :], i[atom, :]

IndexError: list index out of range
```

Curiously the error is a probabilistic function of `n_neurons`.

![p success](https://user-images.githubusercontent.com/5420057/49680914-25a65280-fa68-11e8-8cc8-656fe545a2e0.png)

```python
from collections import defaultdict

import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pandas import DataFrame

import nengo
import nengo_loihi

def attempt(n_neurons):
    with nengo.Network() as model:
        x = nengo.Ensemble(n_neurons, 1)
        nengo.Connection(x.neurons, x, transform=np.ones((x.dimensions, x.n_neurons)))

    try:
        with nengo_loihi.Simulator(model) as sim:
            sim.run(1.0)
    except IndexError:
        return False
    else:
        return True

data = defaultdict(list)
for n_trials in range(50):
    for n_neurons in range(1, 20):
        data['# Neurons'].append(n_neurons)
        data['P(Success)'].append(attempt(n_neurons))

plt.figure()
sns.lineplot(data=DataFrame(data), x='# Neurons', y='P(Success)')
plt.show()
```

**Comment by hunse** [comment]
Likely related to #71 never being figured out/implemented properly.

**Comment by arvoelke** [comment]
I think a suitable work-around is to replace the above connection:

```python
    transform = np.ones((x.dimensions, x.n_neurons))
    nengo.Connection(x.neurons, x, transform=transform)
```

with:

```python
    nengo.Connection(x, x, solver=nengo.solvers.NoSolver(transform.T, weights=False))
```

This seems to work (at least I expect it to do the same thing, and it doesn't throw any exceptions).

In the process of playing around with the above, I noticed something that might help in debugging this error.

If we incorrectly size the transform for `weights=True`---it should be `(n_neurons, n_neurons)` but we pick the following instead---then we get the exact same error in this issue:

```python
    transform = np.ones((x.dimensions, x.n_neurons))  # don't do this
    nengo.Connection(x, x, solver=nengo.solvers.NoSolver(transform, weights=True))
```

**Comment by tbekolay** [comment]
> nengo.Connection(x, x, solver=nengo.solvers.NoSolver(transform.T, weights=False))

Just a caveat for this workaround, my recollection from #85 and #103 is that this works because it moves the weights off-chip. It should do the same-ish thing model output wise, but it does quite a different thing on the chip itself, so it depends on what the user wants.

> If we incorrectly size the transform for `weights=True`---it should be `(n_neurons, n_neurons)` but we pick the following instead---then we get the exact same error in this issue

Interesting, nice detective work ... seems like a bug with weights being transposed somewhere then 👌

**Comment by arvoelke** [comment]
Hmm this is getting a little confusing for me. Is there a way I can access what the splitter has done afterwards, to debug what objects are placed where and which connections are crossing the host-chip boundary? Likewise for where interneurons are being placed (#97)?

I think this doesn't do quite what I wanted either way, because interneurons are placed in between. I was hoping that `Ensemble --(NoSolver)-> Node -> Ensemble` would all be done on-chip using an all-to-all weight matrix (by collapsing the passthrough) but I'm having trouble figuring out what's actually happening. I think it's staying in state-space because of the post encoders, and implementing that by adding interneurons.

Basically.. all I'm trying to do is create an all-to-all recurrent connection, with `n_neurons` exceeding the per-core maximum memory (i.e., `n_neurons>=342`). This has required partitioning the ensemble into multiple sub-ensembles, and solving for the decoders in advance. Is the last step to pre-specify all of the encoders, explicitly multiply these through into a full weight-matrix, and then extract each block to specify a `transform` on each `subensemble.neurons -> subensemble.neurons` connection? Or is there an easier way? If this question is more suitable for the forum, the thread is here: https://forum.nengo.ai/t/how-many-neurons-can-be-fully-connected/706

**Comment by drasmuss** [comment]
Fixed in https://github.com/nengo/nengo-loihi/pull/156
